// @selfxyz/core/dist/common/src/utils/passports/passport_parsing/parseDscCertificateData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseDscCertificateData = parseDscCertificateData;
const brutForceDscSignature_1 = require("./brutForceDscSignature");
const parseCertificateSimple_1 = require("../../certificate_parsing/parseCertificateSimple");
const csca_1 = require("../../csca");
const parsePassportData_1 = require("./parsePassportData");
function parseDscCertificateData(dscCert) {
    let csca, cscaParsed, cscaHashAlgorithm, cscaSignatureAlgorithm, cscaCurveOrExponent, cscaSignatureAlgorithmBits, cscaSaltLength;
    let cscaFound = false;
    if (dscCert.authorityKeyIdentifier) {
        try {
            csca = (0, csca_1.getCSCAFromSKI)(dscCert.authorityKeyIdentifier, true);
            if (csca) {
                cscaParsed = (0, parseCertificateSimple_1.parseCertificateSimple)(csca);
                const details = (0, brutForceDscSignature_1.brutforceSignatureAlgorithmDsc)(dscCert, cscaParsed);
                cscaFound = true;
                cscaHashAlgorithm = details.hashAlgorithm;
                cscaSignatureAlgorithm = details.signatureAlgorithm;
                cscaCurveOrExponent = (0, parsePassportData_1.getCurveOrExponent)(cscaParsed);
                cscaSignatureAlgorithmBits = parseInt(cscaParsed.publicKeyDetails.bits);
                cscaSaltLength = details.saltLength;
            }
        }
        catch (error) { }
    }
    else {
        console.log('js: dscCert.authorityKeyIdentifier not found');
    }
    return {
        cscaFound: cscaFound,
        cscaHashAlgorithm: cscaHashAlgorithm,
        cscaSignatureAlgorithm: cscaSignatureAlgorithm,
        cscaCurveOrExponent: cscaCurveOrExponent,
        cscaSignatureAlgorithmBits: cscaSignatureAlgorithmBits,
        cscaSaltLength: cscaSaltLength,
        csca: csca,
        cscaParsed: cscaParsed,
        cscaBits: cscaSignatureAlgorithmBits,
    };
}
